resources/todomvc/architecture-examples/angular/src/app/todos.service.ts

```diff
@@ -0,0 +1,55 @@
+import { Injectable } from "@angular/core";
+import { Todo } from "./todo";
+
+function uuid() {
+    let uuid = "";
+    for (let i = 0; i < 32; i++) {
+        const random = (Math.random() * 16) | 0;
+        if (i === 8 || i === 12 || i === 16 || i === 20)
+            uuid += "-";
+
+        uuid += (i === 12 ? 4 : i === 16 ? (random & 3) | 8 : random).toString(16);
+    }
+    return uuid;
+}
+
+@Injectable({
+    providedIn: "root",
+})
+export class TodosService {
+    todos: Todo[] = [];
+
+    addItem(title: string): void {
+        const todo: Todo = {
+            id: uuid(),
+            title,
+            completed: false,
+        };
+
+        this.todos = this.todos.concat(todo);
+    }
+
+    removeItem(todo: Todo): void {
+        const index = this.todos.findIndex((t) => t.id === todo.id);
+        this.todos.splice(index, 1);
+    }
+
+    clearCompleted(): void {
+        this.todos = this.todos.filter((todo) => !todo.completed);
+    }
+
+    toggleAll(completed: boolean): void {
+        this.todos = this.todos.map((todo) => ({ ...todo, completed }));
+    }
+
+    getItems(type = "all"): Todo[] {
```

Comment:
I was curious, so I added a log here, and I noticed that this is called 14 times (!!) with  very often (for example on blur, or for each typed character in the header).
Can you please check if there's a way of _not_ doing that? Is there something native to do it only when needed, or maybe can we cache the result, invalidating the cache when some changes are made to the todos array?

Reply:
I think this would be a great use of a pipe (https://angular.io/guide/pipes), which AFAIU is similar to Redux selectors.
However my understanding is that with pipes we either need to go back at the immutable way of doing things: use map, filter, etc, (in that case please disregard my other comments above + change the `removeItem`) or mark the pipe as "impure", and in that case it would be evaluated once for every keystroke and other events (which would still be an improvement over the current situation!).

Tell me what you think :-)

Reply:
I'll take a look - thanks for surfacing. 
At this point I rather fix what we have though instead of refactoring again.

Reply:
@julienw what do you want me to try? 

I did test out the pipe idea, but didn't see any noticeable performance benefit. 
Here is the test branch I was using: 
https://github.com/flashdesignory/Speedometer/blob/todomvc-angular-test/resources/todomvc/architecture-examples/angular/src/app/todos.pipe.ts

Reply:
Thanks for the pointer! I'll have a look tomorrow, to see if there's anything obvious.
Otherwise I wanted to implement some cache like mentioned in the previous comment. I can also have a look at it of you want, see if this changes things.

Reply:
Oh absolutely - it's a team effort and I welcome direct contributions 😄

Reply:
With your code, the pipe was still running as often as before. But I made it work properly by replacing `getTodos` to return `this.todos` directly instead of slicing it (in `todos.service`).

But... this made no difference to the benchmark result. My guess is that we don't exercise this code much in the benchmark as we assign the values directly to the `value` attribute, so the fact that it runs for each keystroke doesn't matter.
So even with this change, like you I see that the pipe implementation (along with using immutable structures) is slower than the current code.

I also tried to implement a cache "manually" but the fact that we change the `completed` value directly on the `todo` item makes things a bit messy.

All in all I think we went through all paths now and I'm happy about landing this version.